app/lib/core/services/authentication/auth_service.dart

```diff
@@ -1,38 +1,108 @@
 import 'dart:async';
+import 'package:app/core/models/user_info.dart';
+import 'package:app/core/services/firestore/firestore_service.dart';
+import 'package:app/core/services/service_locator.dart';
 import 'package:firebase_auth/firebase_auth.dart';
 import 'package:flutter/material.dart';
 
 class AuthenticationService {
   final FirebaseAuth _firebaseAuth = FirebaseAuth.instance;
+  final FirestoreService _firestoreService = locator<FirestoreService>();
 
+  UserInformation _currentUser;
+  UserInformation get currentUser => _currentUser;
+
+  /*
+   This needs to be a dynamic type as we may expect bool or a string as return type.
+   Note: Same logic applies to other methods here.
+  [FirebaseAuthException] - Will return String if something went wrong
+  */
   Future signIn({
     @required String email,
     @required String password,
   }) async {
     try {
-      var user = await _firebaseAuth.signInWithEmailAndPassword(
+      UserCredential authResult =
+          await _firebaseAuth.signInWithEmailAndPassword(
         email: email,
         password: password,
       );
-      return user != null;
-    } catch (e) {
-      return e.message;
+      var user = await _firestoreService.fetchUser(authResult.user.uid);
+      var isUserExist = user != null;
+
+      await _populateCurrentUser(authResult.user);
+
+      print('${user.firstName} currently signed in!');
+      return isUserExist;
+    } on FirebaseAuthException catch (e) {
+      return e.code;
+    }
+  }
+
+  Future signUp({
+    @required String firstName,
+    @required String lastName,
+    @required String userEmailAddress,
+    @required String userPhoneNumber,
+    @required String password,
+  }) async {
+    try {
+      UserCredential authResult =
+          await _firebaseAuth.createUserWithEmailAndPassword(
+        email: userEmailAddress,
+        password: password,
+      );
+
+      // Uses the service to create a new user to Firestore
+      await _firestoreService.addUserToFirestore(UserInformation(
+        userId: authResult.user.uid,
+        firstName: firstName,
+        lastName: lastName,
+        userEmailAddress: userEmailAddress,
+        userPhoneNumber: userPhoneNumber,
+      ));
+
+      return authResult.user != null;
```

Comment:
Instead of returning all these different things, let's make a controlled return type. If you've heard of `Try` and `Failure`, this should be familiar. How about:

- [ ] Add the following and use them

```
enum ResponseType { success, failure, catastrophic_failure }

class Response<T> { 
    final ResponseType type;
    final String reason;
    final Optional<T> response;

    Response.failure({@required this.type, @required this.reason}) :
    response = Optional.empty();

    Response.success({@required this.response}):
    type = ResponseType.success,
    reason = "Successful";
}
```
This will help us to have very clean and thorough error handling for future work.

Reply:
Will implement a better error handling in my next commit